### ISSUE TITLE
Ignore all `#[test]`-related code automatically

### ISSUE THREAD

**Issue by Swatinem** [taiki-e/cargo-llvm-cov#123]
UPDATE3(taiki-e):

> If you want to ignore all `#[test]`-related code, you can use module-level `#[coverage(off)]` attribute:
>
> ```rust
> #![cfg_attr(coverage_nightly, feature(coverage_attribute))]
>
> #[cfg(test)]
> #[cfg_attr(coverage_nightly, coverage(off))]
> mod tests {
>     // ...
> }
> ```
> 
> cargo-llvm-cov excludes code contained in the directory named `tests` and file named `tests.rs`/`*_tests.rs`/`*-tests.rs` from the report by default, so you can also use it instead of `#[coverage(off)]` attribute.

~~`#[coverage(off)]` attribute has been stabilized in https://github.com/rust-lang/rust/pull/130766 (will be included in Rust 1.85).~~ UPDATE: stabilization has been reverted in https://github.com/rust-lang/rust/pull/134672.

---

UPDATE2(taiki-e):

See https://github.com/taiki-e/cargo-llvm-cov/issues/123#issuecomment-2423715856

---

UPDATE(taiki-e):

> If you want to ignore all `#[test]`-related code, consider using [coverage-helper](https://github.com/taiki-e/coverage-helper) crate version 0.2+.
>
> cargo-llvm-cov excludes code contained in the directory named `tests` from the report by default, so you can also use it instead of coverage-helper crate.

https://github.com/taiki-e/cargo-llvm-cov#exclude-function-from-coverage

---

AFAIK this is not currently possible (though I would love to be proven wrong), so just filing this as a nice-to-have tracking issue.

What I would like is that only library code counts towards coverage. Having all of the `#[test]` and `#[cfg(test)]` code (fn and whole utility modules) count towards coverage metrics right now skews those numbers up quite a bit.

Right now `cargo-llvm-cov` automatically ignores all the `examples` and `tests` code via `-ignore-filename-regex` but unit test functions are still counted (and possibly doctests once those are being fixed [tentative fix in #122]).

**Comment by taiki-e** [comment]
The idea that probably works here is a helper attribute that generates `#[cfg_attr(coverage, no_coverage)]`.

```rs
#[cfg(coverage)]
use coverage_helper::test;

#[test]
fn test() {
}
```

generated:

```rs
#[cfg(coverage)]
use coverage_helper::test;

#[cfg_attr(coverage, no_coverage)]
#[::core::prelude::v1::test]
fn test() {
}
```

**Comment by scouten-adobe** [comment]
@taiki-e I'm not an expert on attribute macros. Would you (or anyone) be kind enough to translate this into code that we could implement?

Adding the `no_coverage` attribute in front of every test seems unsustainable.

**Comment by taiki-e** [comment]
Above my approach is now implemented as https://github.com/taiki-e/coverage-helper.

**Comment by scouten-adobe** [comment]
I appreciate the coverage-helper crate, but since the [`#[no_coverage]` attribute](https://github.com/rust-lang/rust/issues/84605) is not yet stabilized, I've decided not to use this approach in my project.

For those following along, if you want to exclude test-related code from coverage and base your CI builds on stable Rust versions, there's a more drastic approach that does seem to solve the problem. See https://github.com/adobe/xmp-toolkit-rs/pull/66.

**Comment by mati865** [comment]
There is upstream PR with the fix https://github.com/rust-lang/rust/pull/120183

**Comment by taiki-e** [comment]
> There is an upstream PR with the fix [rust-lang/rust#120183](https://github.com/rust-lang/rust/pull/120183)

Hmm, I thought that PR was regarding the closure generated internally by the test/bench attributes and not for the test/bench code itself.

**Comment by Zalathar** [comment]
Yeah, that PR only gets rid of an implementation-detail closure that happens to be emitted by the `#[test]` macro. It has no effect on whether the user’s actual test function is instrumented or not.

**Comment by taiki-e** [comment]
I'm surprised myself that I (probably) haven't documented this, but a way that doesn't depend on the unstable features that have worked since the earliest versions is to put unit tests in a directory named `tests`.

My projects over the past few years have adopted a directory structure that obviously has this in mind. 

- https://github.com/taiki-e/portable-atomic/tree/main/src/tests
- https://github.com/taiki-e/parse-changelog/tree/main/src/tests
- https://github.com/taiki-e/dependabot-config/tree/main/src/tests
- https://github.com/taiki-e/atomic-maybe-uninit/tree/main/src/tests

**Comment by wyfo** [comment]
Would it make sense to exclude `tests.rs` files also? If the test module is small, user may prefer having `src/tests.rs` instead of `src/tests/mod.rs`.

**Comment by taiki-e** [comment]
> exclude `tests.rs` files

Seems reasonable to me.

**Comment by taiki-e** [comment]
In the recent rustc, it seems the module-level `coverage(off)` is supported:

```rs
#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))] // <---
mod tests {
   // ...
}
```

(Perhaps coverage-helper library should be deprecated for most cases in favor of the module-level `coverage(off)` ...)

**Comment by Zalathar** [comment]
Oh yes, that was https://github.com/rust-lang/rust/pull/126721.

**Comment by taiki-e** [comment]
`#[coverage(off)]` attribute has been stabilized in https://github.com/rust-lang/rust/pull/130766 (will be included in Rust 1.85).

And I've updated documentation to recommend the pattern described in https://github.com/taiki-e/cargo-llvm-cov/issues/123#issuecomment-2423715856:

> If you want to ignore all `#[test]`-related code, you can use module-level `#[coverage(off)]` attribute:
>
> ```rust
> #[cfg(test)]
> #[cfg_attr(coverage, coverage(off))]
> mod tests {
>     // ...
> }
> ```
> 
> cargo-llvm-cov excludes code contained in the directory named `tests` from the report by default, so you can also use it instead of `#[coverage(off)]` attribute.
